### PR TITLE
Convert form state to use state 

### LINKED ISSUE #2631: Nested useFieldArray elements seem to keep default values after reset is called
**Describe the bug**
If default data is set and an attempt to reset data with a new value. useFieldArray seems to return the defaulted value rather than the reset one

**To Reproduce**

1. create a form with nested useFieldArrays with a defaultValue set
2. At the form root create a useEffect and a setTimeout to set the default data via reset.
3. You still see the original defaulted value

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-tctqs?file=/src/index.js

**Expected behavior**
useFieldArray value should be the new value not the default

**Desktop (please complete the following information):**

OS: Mac (Mojave 10.14.6)
Browser: Chrome
Version: 84

### LINKED ISSUE #2655: Disabled input values appear in handleSubmit data.
**Describe the bug**
Values for disabled inputs still appear in `handleSubmit` data.

**To Reproduce**

1. Create a form with a disabled input.
2. Submit the form and inspect the data passed to `handleSubmit`.
3. See the value for the disabled input.

**Codesandbox link (Required)**
https://codesandbox.io/s/sweet-morning-dzc1f

**Expected behavior**
Disabled input values do not appear in `handleSubmit` data.

**Desktop (please complete the following information):**

- OS: macOS
- Browser: Chrome
- Version latest

**Additional context**
https://github.com/react-hook-form/react-hook-form/discussions/2654

### PR DESCRIPTION
**Pull request by bluebill1049**
## 🎉 Summary

- convert `useRef` state to `useState`
- export `errors` object inside `formState` and keep `errors` at the root level as an alias
- update unit tests accordingly 
- update cypress automation tests
- convert `validatedFields` & `fieldsWithvalidation` to Object type
- upgrade TS4
- upgrade Cypress 5
- improve Cypress integration tests with state update by @pmaier983 
- minor Type improvement.

## 🐞 Bugs

- close #2631 reset with useFieldArray
- close fix #2655 disable input

let me know if you have any questions or concerns.

### CODE REVIEW

**Comment by bluebill1049** [comment]
I am still working on it, but feel free to give early feedback @stramel @adhrinae I will amend accordingly.

File: cypress/integration/conditionalField.ts [hunk 1]
```diff
@@ -2,52 +2,52 @@ describe('ConditionalField', () => {
...
-      '{"dirtyFields":[],"isSubmitted":false,"submitCount":0,"touched":[],"isDirty":false,"isSubmitting":false,"isValid":false}',
+      '{"isDirty":false,"dirtyFields":[],"isSubmitted":false,"submitCount":0,"touched":[],"isSubmitting":false,"isValid":false}',
     );
 
     cy.get('select[name="selectNumber"]').select('1');
     cy.get('input[name="firstName"]').type('bill');
     cy.get('input[name="lastName"]').type('luo');
     cy.get('input[name="lastName"]').blur();
     cy.get('#state').contains(
-      '{"dirtyFields":["selectNumber","firstName","lastName"],"isSubmitted":false,"submitCount":0,"touched":["selectNumber","firstName","lastName"],"isDirty":true,"isSubmitting":false,"isValid":true}',
```
Comment (pmaier983) [inline cypress/integration/conditionalField.ts:13]: Would it be better to change this into an object and do a deep equality check using lodash (would be in devDeps ofc.) or some home baked method? It would prevent these small diffs.

Comment (bluebill1049) [reply cypress/integration/conditionalField.ts:13]: sounds good to me! feel free to commit on top this branch 🙏

Comment (pmaier983) [reply cypress/integration/conditionalField.ts:13]: I was unable to commit to this branch directly :/ It had some bugs i was unable to fix in its current state. Sorry Bill!

**Comment by bluebill1049** [comment]
Hi guys, Could you please take a detailed review of this one? many thanks. 🙏

File: src/useForm.ts [hunk 2]
```diff
@@ -141,15 +144,26 @@ export function useForm<
...
     current: { isOnBlur: isReValidateOnBlur, isOnChange: isReValidateOnChange },
   } = React.useRef(modeChecker(reValidateMode));
+
   contextRef.current = context;
   resolverRef.current = resolver;
-
-  const reRender = React.useCallback(
-    () => !isUnMount.current && render({}),
+  formStateRef.current = formState;
+  errorsRef.current = formState.errors;
```
Comment (keiya01) [inline src/useForm.ts:157]: Is `errorsRef` unused ?

File: src/useForm.ts [hunk 3]
```diff
@@ -504,16 +557,17 @@ export function useForm<
...
             error = (get(errors, name)
               ? { [name]: get(errors, name) }
               : {}) as FlatFieldErrors<TFieldValues>;
 
-            if (previousFormIsValid !== isValidRef.current) {
+            if (previousFormIsValid !== isEmptyObject(errors)) {
               shouldRender = true;
             }
+
+            isValid = isEmptyObject(errors);
```
Comment (keiya01) [inline src/useForm.ts:570]: I think these line can be changed as below.

```js
isValid = isEmptyObject(errors);
if(previousFormIsValid !== isValid) {
  shouldRender = true;
}
```

**Review by keiya01 — COMMENTED** [review]
I reviewed the PR. 👀

File: src/useForm.ts [hunk 4]
```diff
@@ -127,12 +121,21 @@ export function useForm<
...
+  const [formState, setFormState] = React.useState<FormState<TFieldValues>>({
+    isDirty: false,
+    dirtyFields: {},
+    isSubmitted: false,
+    submitCount: 0,
+    touched: {},
+    isSubmitting: false,
+    isValid: !isOnSubmit,
+    errors: {},
+  });
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.ts:138]: Thanks @keiya01 :hear

File: src/useForm.ts (hunk 2 above)
Comment (bluebill1049) [reply src/useForm.ts:157]: oh wow! your 👀 🙄

File: src/useForm.ts (hunk 3 above)
Comment (bluebill1049) [reply src/useForm.ts:570]: oh yeah!

**Comment by bluebill1049** [comment]
Thanks very much, @keiya01 for your review. Let's start to use your package after this one is approved by everyone.

File: src/useForm.ts [hunk 5]
```diff
@@ -546,13 +595,13 @@ export function useForm<
...
-      return payload.reduce(
-        (previous, name) => ({
-          ...previous,
-          [name]: getFieldValue(fieldsRef, name, unmountFieldsStateRef),
-        }),
-        {},
-      );
+      const data = {};
+
+      for (const name of payload) {
```
Comment (pmaier983) [inline src/useForm.ts:600]: I think I liked the old way using reduce (just a bit more functional). Why the change?

File: src/useForm.ts [hunk 6]
```diff
@@ -1196,26 +1183,47 @@ export function useForm<
...
     control,
+    formState: isProxyEnabled
+      ? new Proxy<FormStateProxy<TFieldValues>>(formState, {
+          get: (obj, prop: keyof FormStateProxy) => {
+            if (
+              process.env.NODE_ENV !== 'production' &&
+              prop === 'isValid' &&
+              isOnSubmit
+            ) {
+              console.warn(
```
Comment (pmaier983) [inline src/useForm.ts:1208]: I love the doc link with the warning, May open a PR to make this happen everywhere.

**Review by pmaier983 — COMMENTED** [review]
I am not 100% sure I understand the benefits of the formState transition (besides a simpler codebase), so i don't have any high level comments. Mostly just specific in the weeds code comments. Looks like a nice simplifier though! (and i am finally dipping my toe into the core)

File: src/__mocks__/reconfigureControl.ts [hunk 7]
```diff
@@ -55,14 +50,17 @@ export const reconfigureControl = (
...
-    submitCount: 0,
-    touched: {},
-    isSubmitting: false,
-    isValid: false,
+  formStateRef: {
+    current: {
+      errors: {},
+      isDirty: false,
+      isSubmitted: false,
+      dirtyFields: {},
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/__mocks__/reconfigureControl.ts:58]: It's a tech debt, we should convert them all into `object`, for consistency. We will get to it.

File: src/logic/getFieldsValues.ts [hunk 8]
```diff
@@ -32,7 +32,7 @@ export default <TFieldValues extends FieldValues>(
   }
 
   return deepMerge(
-    transformToNestObject((unmountFieldsStateRef || {}).current || {}),
+    { ...((unmountFieldsStateRef || {}).current || {}) },
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/logic/getFieldsValues.ts:35]: `deepMerge` is mutable. it's safer to do a shallow clone there.

File: src/useFieldArray.ts [hunk 9]
```diff
@@ -187,26 +166,25 @@ export const useFieldArray = <
       readFormStateRef.current.dirtyFields ||
       readFormStateRef.current.isDirty
     ) {
-      set(dirtyFieldsRef.current, name, [
-        ...(get(dirtyFieldsRef.current, name) ||
-          fillEmptyArray(fields.slice(0, 1))),
+      set(dirtyFields, name, [
+        ...(get(dirtyFields, name) || fillEmptyArray(fields.slice(0, 1))),
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useFieldArray.ts:170]: wow! you are 100% right! let's delete some code. 👯

File: src/useFieldArray.ts [hunk 10]
```diff
@@ -216,92 +194,63 @@ export const useFieldArray = <
...
-      set(
-        dirtyFieldsRef.current,
-        name,
-        removeArrayAt(get(dirtyFieldsRef.current, name), index),
-      );
+      set(dirtyFields, name, removeArrayAt(get(dirtyFields, name), index));
 
-      if (!unique(get(dirtyFieldsRef.current, name, [])).length) {
-        unset(dirtyFieldsRef.current, name);
+      if (!unique(get(dirtyFields, name, [])).length) {
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useFieldArray.ts:251]: yes, please! PRs always welcome, you can commit into this branch as well :)

File: src/useForm.test.tsx [hunk 11]
```diff
@@ -255,19 +258,23 @@ describe('useForm', () => {
...
-    it('should not call findRemovedFieldAndRemoveListener when field variable does not exist', () => {
+    it('should not call findRemovedFieldAndRemoveListener when field variable does not exist', async () => {
       const mockListener = jest.spyOn(
         findRemovedFieldAndRemoveListener,
         'default',
       );
       const { result } = renderHook(() => useForm());
 
-      result.current.unregister('test');
+      await act(async () => {
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.test.tsx:275]: I am not 100% sure on this one, I was trying to get rid of testing lib warning, maybe @keiya01 have more clues.

File: src/useForm.ts [hunk 12]
```diff
@@ -158,38 +170,51 @@ export function useForm<
...
 
-      if (shouldReRender && !isNullOrUndefined(shouldRender)) {
-        reRender();
-        return true;
+      if (shouldReRender || !isEmptyObject(state)) {
+        updateFormState({
+          ...state,
+          errors: formStateRef.current.errors,
+          isValid: resolverRef.current
+            ? !!isValid
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.ts:213]: you are right! you should use the suggestion feature! then I can commit right away.

File: src/useForm.ts (hunk 5 above)
Comment (bluebill1049) [reply src/useForm.ts:600]: I prefer the old way as well, but this way you can set up a nested object. the previous version is just a flat object.

File: src/useForm.ts (hunk 6 above)
Comment (bluebill1049) [reply src/useForm.ts:1208]: send more PRs 🙏🏻

File: src/useForm.ts [hunk 13]
```diff
@@ -471,31 +513,40 @@ export function useForm<
...
+              touched: formStateRef.current.touched,
+            };
           }
 
           if (shouldSkipValidation) {
             renderWatchedInputs(name);
-            return shouldRender && reRender();
+            return (
+              (!isEmptyObject(state) ||
+                (shouldRender && isEmptyObject(state))) &&
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.ts:547]: because `setDirty` function just update the dirty value, so we don't return anything, now we have to return `state` wether is `{}` or `{dirty: true, dirtyFields: [....]}`, the logic hurt my eyes as well... but i give up it's working, feel free to suggest better.

[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.ts:547]: because `setDirty` function just update the dirty value, so we don't return anything, now we have to return `state` wether is `{}` or `{dirty: true, dirtyFields: [....]}`, the logic hurt my eyes as well... but i give up it's working, feel free to suggest better.

File: src/useForm.ts (hunk 5 above)
Comment (bluebill1049) [reply src/useForm.ts:600]: also a little big faster compare with .reduce

File: src/__mocks__/reconfigureControl.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/__mocks__/reconfigureControl.ts:58]: actually i will make a PR on top of this, hopefully it's not too hairy.

File: src/useForm.test.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (keiya01) [reply src/useForm.test.tsx:275]: I think we can use `waitFor` as well but `waitFor` is a little difficult to use correctly. Because `waitFor` will wait for stopping to throw error. So we might write wrong test like bellow when we want to test async function.

```jsx
const Component = () => {
  const { register } = useForm({ reset: false });
  return (
    <form onSubmit={handleSubmit(() => {})}>
      <input ref={register({ minLength: 5 })}/>
      <button>submit</button>
    </form>
  );
};

test('should not reset when form has error', async () => {
  render(<Component/>);
  fireEvent.input(screen.getByRole('textbox'), { target: { value: 'test' } });
  fireEvent.submit(screen.getByRole('form'));
  // This case will pass. But submit have not been completed yet because submit event is async.
  await   waitFor(() => expect(screen.getByRole('textbox')).toHaveValue('test');
})
```
Therefore I prefer to use `await act(async () => { ... })` rather than to use `waitFor`.
But we can avoid the wrong code by paying attention.

File: src/useForm.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useForm.ts:213]: actually !!isValid needed because isValid is optional.

File: src/useFieldArray.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useFieldArray.ts:170]: actually we do need `[]` as well :)

File: src/__mocks__/reconfigureControl.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/__mocks__/reconfigureControl.ts:58]: `validFields` & `fieldsWithValidation` has been addressed.

**Comment by stramel** [comment]
Looks like the net size dropped 300KB across all bundles and functionality is essentially the same. I'm good with this as long as the checks are all passing. I don't see any reason that this should be an issue.

**Review by stramel — DISMISSED** [review]

**Comment by bluebill1049** [comment]
> Looks like the net size dropped 300KB across all bundles and functionality is essentially the same. I'm good with this as long as the checks are all passing. I don't see any reason that this should be an issue.

It's actually dropped by refactoring two data types, but yeah thanks for the review @stramel 👍

**Comment by bluebill1049** [comment]
fixed @pmaier983

**Review by pmaier983 — DISMISSED** [review]
Looks Awesome! Really great work bill!

File: src/useFieldArray.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (pmaier983) [reply src/useFieldArray.ts:170]: When you say this, what do you mean? did `[null]` not work?

**Comment by bluebill1049** [comment]
> Looks Awesome! Really great work bill!

you too :)

File: cypress/integration/controller.ts [hunk 14]
```diff
@@ -50,30 +50,30 @@ describe('controller basic form validation', () => {
...
 
-    cy.get('p').should('have.length', 1);
-    cy.get('#renderCount').contains('8');
+    cy.get('p').should('have.length', 0);
+    cy.get('#renderCount').contains('9');
   });
 
-  // it('should validate the form with onChange mode and reset the form', () => {
-  //   cy.visit('http://localhost:3000/controller/onChange');
+  it('should validate the form with onChange mode and reset the form', () => {
```
Comment (pmaier983) [inline cypress/integration/controller.ts]: Thank you for fixing! I wonder why it was broken in the first place!

Comment (bluebill1049) [reply cypress/integration/controller.ts]: no idea :(

**Comment by bluebill1049** [comment]
plan to merge it this Friday 👍

File: src/useFieldArray.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bluebill1049) [reply src/useFieldArray.ts:169]: `[null]` not work, because we need `[]`.
